"""Client gRPC pour tester le service d'urgence."""
import grpc
import emergency_pb2
import emergency_pb2_grpc


def run():
    """Ex√©cute les tests du client gRPC."""
    # Connexion au serveur
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = emergency_pb2_grpc.EmergencyServiceStub(channel)
        
        print("üöë Tests du Service gRPC - Urgences")
        print("=" * 60)
        
        # Test 1: R√©cup√©rer tous les v√©hicules
        print("\n1Ô∏è‚É£ Test: GetAllVehicles()")
        print("-" * 60)
        try:
            response = stub.GetAllVehicles(emergency_pb2.Empty())
            print(f"‚úÖ Nombre de v√©hicules: {len(response.vehicles)}")
            for vehicle in response.vehicles[:3]:  # Afficher les 3 premiers
                print(f"   ‚Ä¢ {vehicle.identifier} ({vehicle.vehicle_type}) - {vehicle.status}")
                print(f"     Station: {vehicle.station}")
        except grpc.RpcError as e:
            print(f"‚ùå Erreur: {e.details()}")
        
        # Test 2: R√©cup√©rer un v√©hicule par ID
        print("\n2Ô∏è‚É£ Test: GetVehicle(id=1)")
        print("-" * 60)
        try:
            response = stub.GetVehicle(emergency_pb2.VehicleRequest(id=1))
            print(f"‚úÖ V√©hicule trouv√©:")
            print(f"   ID: {response.id}")
            print(f"   Identifiant: {response.identifier}")
            print(f"   Type: {response.vehicle_type}")
            print(f"   Statut: {response.status}")
            print(f"   Station: {response.station}")
            print(f"   √âquipage: {response.crew_size} personnes")
            print(f"   Position: ({response.latitude}, {response.longitude})")
        except grpc.RpcError as e:
            print(f"‚ùå Erreur: {e.details()}")
        
        # Test 3: R√©cup√©rer les ambulances disponibles
        print("\n3Ô∏è‚É£ Test: GetAvailableVehicles(type='ambulance')")
        print("-" * 60)
        try:
            response = stub.GetAvailableVehicles(
                emergency_pb2.VehicleTypeRequest(vehicle_type="ambulance")
            )
            print(f"‚úÖ Ambulances disponibles: {len(response.vehicles)}")
            for vehicle in response.vehicles:
                print(f"   ‚Ä¢ {vehicle.identifier} - {vehicle.station}")
        except grpc.RpcError as e:
            print(f"‚ùå Erreur: {e.details()}")
        
        # Test 4: R√©cup√©rer les interventions actives
        print("\n4Ô∏è‚É£ Test: GetActiveInterventions()")
        print("-" * 60)
        try:
            response = stub.GetActiveInterventions(emergency_pb2.Empty())
            print(f"‚úÖ Interventions actives: {len(response.interventions)}")
            for intervention in response.interventions:
                print(f"   ‚Ä¢ #{intervention.id} - {intervention.intervention_type.upper()}")
                print(f"     Priorit√©: {intervention.priority}")
                print(f"     Adresse: {intervention.address}")
                print(f"     Statut: {intervention.status}")
                if intervention.assigned_vehicle_id:
                    print(f"     V√©hicule assign√©: ID {intervention.assigned_vehicle_id}")
        except grpc.RpcError as e:
            print(f"‚ùå Erreur: {e.details()}")
        
        # Test 5: Cr√©er une nouvelle intervention
        print("\n5Ô∏è‚É£ Test: CreateIntervention()")
        print("-" * 60)
        try:
            new_intervention = emergency_pb2.InterventionInput(
                intervention_type="medical",
                priority="high",
                address="10 Rue de la R√©publique, 75001 Paris",
                latitude=48.8600,
                longitude=2.3400,
                assigned_vehicle_id=1,
                description="Chute avec suspicion de fracture"
            )
            response = stub.CreateIntervention(new_intervention)
            print(f"‚úÖ Intervention cr√©√©e:")
            print(f"   ID: {response.id}")
            print(f"   Type: {response.intervention_type}")
            print(f"   Priorit√©: {response.priority}")
            print(f"   Adresse: {response.address}")
            print(f"   V√©hicule assign√©: {response.assigned_vehicle_id}")
        except grpc.RpcError as e:
            print(f"‚ùå Erreur: {e.details()}")
        
        # Test 6: Mettre √† jour le statut d'un v√©hicule
        print("\n6Ô∏è‚É£ Test: UpdateVehicleStatus()")
        print("-" * 60)
        try:
            status_update = emergency_pb2.StatusUpdate(
                vehicle_id=1,
                new_status="on_mission",
                latitude=48.8610,
                longitude=2.3410
            )
            response = stub.UpdateVehicleStatus(status_update)
            print(f"‚úÖ Statut du v√©hicule mis √† jour:")
            print(f"   {response.identifier}: {response.status}")
            print(f"   Nouvelle position: ({response.latitude}, {response.longitude})")
        except grpc.RpcError as e:
            print(f"‚ùå Erreur: {e.details()}")
        
        print("\n" + "=" * 60)
        print("‚úÖ Tests termin√©s!")


if __name__ == '__main__':
    run()
